### ISSUE TITLE
HutterFeatures observation function

### ISSUE THREAD

**Issue by dchetelat** [ds4dm/ecole#173]
It would be nice to have the instance features from Figure 3 of Hutter et al. (2011), as an observation function. This observation function would be valid during problem creation, and I suppose it could be valid during all presolving as well. It would be a "flat" counterpart to MilpBipartite, in the same way that the Khalil features are a "flat" counterpart to NodeBipartite.

This would be useful for the algorithm configuration challenge in the upcoming Neurips 2021 competition.

![image](https://user-images.githubusercontent.com/8616319/116733323-f175b880-a9b9-11eb-9f8c-09467154658b.png)

_Hutter, F., Hoos, H. H., & Leyton-Brown, K. (2011, January). [Sequential model-based optimization for general algorithm configuration.](https://www.cs.ubc.ca/~hutter/papers/10-TR-SMAC.pdf) In International conference on learning and intelligent optimization (pp. 507-523). Springer, Berlin, Heidelberg._

**Comment by dchetelat** [comment]
We agreed at the meeting that features 35-39 should be excluded. However, the question remains about what to do with the LP features 21-24. I checked the paper again and I saw this:

> The 76 parameters we selected affect all aspects of CPLEX. They include **12 preprocessing parameters** (mostly categorical); 17 MIP strategy parameters (mostly categorical); 11 categorical parameters deciding how aggressively to use which types of cuts; 9 numerical MIP “limits” parameters; 10 simplex parameters (half of them categorical); 6 barrier optimization parameters(mostly categorical); and 11 further parameters.

Since they optimize CPLEX preprocessing parameters, it seems clear that this is not computed "at the root node", but rather before CPLEX is called at all. So the LP solving must be done prior to anything happening. Considering this, I see two possibilities for the features.
1) We just discard the LP relaxation features from the observation function. Simple and easy.
2) When the observation function is called, the model is copied, its variable types are changed for continuous, and the continuous model is solved, yielding the LP relaxation solution, which is used to compute the features.

In possibility 2), of course feature computation time will be much more significant, but I'll note that Hutter et al. themselves don't care about it, because this is only done for them at training time.

> Note that in contrast to per-instance approaches, such as SATzilla [29,27], **instance features are only needed for the training instances**: the end result of algorithm configuration is a single parameter configuration that is used without a need to compute features for test instances. As a corollary, the time required for feature computation is not as crucial in algorithm configuration as it is in per-instance approaches: in per-instance approaches, feature computation has to be counted as part of the time required to solve test instances, **while in algorithm configuration no features are computed for test instances at all**. In fact, features for the training instances may well be the result of an extensive offline analysis of those training instances, or can even be taken from the literature.

So perhaps we shouldn't care about speed either?

**Comment by AntoinePrv** [comment]
I like 2. to be closer to the original.

**Comment by dchetelat** [comment]
Ok perfect, let's do that :+1:

**Comment by dchetelat** [comment]
Note: the features description says that features 12-17 should be the max, min, std, 25% and 75% quantile of the variable node degrees, but that's 5 features, whereas 12-17 should be 6 features. So either they miscounted, or one feature is missing: considering the mean is missing, I find this more likely, so I will include the mean as well.

**Comment by AntoinePrv** [comment]
For 3., [`SCIPgetNNZs`](https://www.scipopt.org/doc-7.0.2/html/group__PublicSolvingStatsMethods.php#gacd0eeb67c7695127c44b32001a2481b2) return the NNZ in the _transformed_ problem and (as I recall) _0_ otherwise...

**Comment by dchetelat** [comment]
Here's some additional info about the features, that could be useful in the implementation.

From Frank Hutter's Ph.D. thesis:

![image](https://user-images.githubusercontent.com/8616319/119882862-e3558200-befc-11eb-977a-78c80550d1bb.png)
![image](https://user-images.githubusercontent.com/8616319/119882915-f6685200-befc-11eb-94ba-551f85c35ada.png)

**Comment by dchetelat** [comment]
From Leyton-Brown et al. (2002):

![image](https://user-images.githubusercontent.com/8616319/119883142-32031c00-befd-11eb-9ffa-dd47bee65d9f.png)
![image](https://user-images.githubusercontent.com/8616319/119883183-3f200b00-befd-11eb-8414-ddabc69a884e.png)
